src/net.c

```diff
@@ -753,6 +759,202 @@ static int hyper_set_interface_mtu(struct rtnl_handle *rth,
 				IFLA_MTU);
 }
 
+static char* hyper_ifreq_to_mac_addr(struct ifreq ifr)
+{
+	char *mac_addr;
+
+	if (asprintf(&mac_addr,
+		     "%02x:%02x:%02x:%02x:%02x:%02x",
+		     (uint8_t)ifr.ifr_hwaddr.sa_data[0],
+		     (uint8_t)ifr.ifr_hwaddr.sa_data[1],
+		     (uint8_t)ifr.ifr_hwaddr.sa_data[2],
+		     (uint8_t)ifr.ifr_hwaddr.sa_data[3],
+		     (uint8_t)ifr.ifr_hwaddr.sa_data[4],
+		     (uint8_t)ifr.ifr_hwaddr.sa_data[5]) == -1) {
+		fprintf(stderr, "failed to allocate mac_addr string\n");
+		return NULL;
+	}
+
+	return mac_addr;
+} 
+
+/*!
+ * Check hardware address related to the provided network interface name
+ * matches the expected hardware address. It expects EUI-48 MAC addresses.
+ *
+ * \param device network interface name.
+ * \param mac_addr hardware address (EUI-48) to match with \p device one.
+ *
+ * \note In case the function succeeds, it returns 0. In case the function
+ * fails, it returns -1.
+ */
+static int hyper_check_device_match_mac_addr(const char *device,
+					     const char *mac_addr)
+{
+	int ret = -1;
+	struct ifreq ifr;
+	int sock;
+	char *tmp_mac_addr;
+
+	if (!device && !strcmp(device, "")) {
+		fprintf(stderr, "invalid device\n");
+		goto err;
+	}
+
+	if (!mac_addr && !strcmp(mac_addr, "")) {
+		fprintf(stderr, "invalid mac_addr\n");
+		goto err;
+	}
+
+	sock = socket(PF_INET, SOCK_DGRAM, 0);
+	if (sock == -1) {
+		fprintf(stderr, "failed to get socket handle: %s\n",
+			strerror(errno));
+		goto err;
+	}
+
+	if (!strncpy(ifr.ifr_name, device, strlen(device) + 1)) {
+		fprintf(stderr, "strncpy failed to copy interface"
+			" name: %s\n", strerror(errno));
+		goto err1;
+	}
+
+	if (ioctl(sock, SIOCGIFHWADDR, &ifr) == -1) {
+		fprintf(stderr, "ioctl SIOCGIFHWADDR failed: %s\n",
+			strerror(errno));
+		goto err1;
+	}
+
+	tmp_mac_addr = hyper_ifreq_to_mac_addr(ifr);
+	if (!tmp_mac_addr) {
+		fprintf(stderr, "tmp_mac_addr is NULL\n");
+		goto err1;
+	}
+
+	if (!strcasecmp(mac_addr, tmp_mac_addr)) {
+		fprintf(stderr, "device mac address found %s does not match"
+			" with the expected mac address %s\n",
+			tmp_mac_addr, mac_addr);
+		goto err2;
+	}
+
+	free(tmp_mac_addr);
+	close(sock);
+	return 0;
+
+err2:
+	free(tmp_mac_addr);
+err1:
+	close(sock);
+err:
+	return ret;
```

Comment:
You don't really need `ret` here as it's always going to be `-1`. Alternatively, you could initialise `sock` and `tmp_mac_addr` which would allow you to simplify these errr handling at the end (by dropping 2 of the labels).